docs/content/guides/dagster/pythonic-resources.mdx

```diff
@@ -0,0 +1,538 @@
+
+To use environment variables, pass an <PyObject object="EnvVar" /> when constructing your resource.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_env_vars endbefore=end_new_resources_env_vars
+from dagster._config.field_utils import EnvVar
+from dagster import Definitions
+from dagster._config.structured_config import ConfigurableResource
+
+
+class CredentialsResource(ConfigurableResource):
+    username: str
+    password: str
+
+
+defs = Definitions(
+    assets=...,
+    resources={
+        "credentials": CredentialsResource(
+            username=EnvVar("MY_USERNAME"),
+            password=EnvVar("MY_PASSWORD"),
+        )
+    },
+)
+```
+
+### Configuring resources at runtime
+
+In some cases, you may want to specify configuration for a resource at runtime, in the launchpad or in a <PyObject object="RunRequest" /> for a [schedule](/concepts/partitions-schedules-sensors/schedules) or [sensor](/concepts/partitions-schedules-sensors/sensors). For example, you may want a sensor-triggered run to specify a different target table in a database resource for each run.
+
+You can use the `configure_at_launch()` method to defer the construction of a configurable resource until runtime.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_runtime endbefore=end_new_resource_runtime
+from dagster._config.structured_config import ConfigurableResource
+from dagster import Definitions, asset
+
+
+class DatabaseResource(ConfigurableResource):
+    table: str
+
+    def read(self):
+        ...
+
+
+@asset
+def data_from_database(db_conn: DatabaseResource):
+    return db_conn.read()
+
+
+defs = Definitions(
+    assets=[data_from_database],
+    resources={"db_conn": DatabaseResource.configure_at_launch()},
+)
+```
+
+#### Configuring a resource at runtime in Python code
+
+Then, configuration for the resource can be provided at runtime in the launchpad or in Python code using the `config` parameter of the <PyObject object="RunRequest" />:
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_runtime_launch endbefore=end_new_resource_runtime_launch
+from dagster import sensor, define_asset_job, RunRequest
+
+update_data_job = define_asset_job(name="update_data_job", selection=[data_from_database])
+
+
+@sensor(job=update_data_job)
+def table_update_sensor():
+    tables = ...
+    for table_name in tables:
+        yield RunRequest(
+            run_config=RunConfig(
+                resources={
+                    "db_conn": DatabaseResource(table=table_name),
+                },
+            ),
+        )
+```
```

Comment:
Wanted to flag this this will potentially be confusing because the behavior here will be slightly different than passing resources to `Definitions`, in particular when raw resource objects are in play. I anticipate that people will copy and paste resource dictionaries around alternatively pass them to Definitions and this. There will be slight differences in behavior.